P2553 implementation

crtrott opened the pull request.

crtrott committed:
Update cuda stencil benchmark to choose index type

crtrott committed:
Add option for using user index type in offset compute

This adds an internal macro, if defined the offset calculation
of layout_left/right will cast the extents.extent to the
integer type which was used to index into the mapping.

Remove the casting inside of mdspan itself.

crtrott committed:
Enable the storage type of extents to be defined

This doesn't extent the API but it allows an implementation
detail level overwrite of what extents use. Mostly we
want to figure out the perf consequences of choosing different
index types.

crtrott committed:
Update stencil benchmark to allow exploration of different index types

crtrott committed:
P2553: make extents templated on size_type

This changes only the extents class

crtrott committed:
P2553: fix all other mdspan classes using extents

I.e. hand on/use/etc. size_type template parameter

crtrott committed:
P2553: fix tests

dalg24 commented on include/experimental/__p0009_bits/layout_left.hpp:
```diff
@@ -57,6 +57,8 @@ struct layout_stride;
   template <class Extents>
   class mapping {
+  public:
+    using size_type = typename Extents::size_type;
```
Plz move all the public type member definitions here (`layout_type`, `extents_type`)

crtrott replied:
will do later

crtrott committed:
P2553: Fix it all up

Note we need to rewrite the history and do some more cleanup.
But now everything works.

dalg24 commented on benchmarks/copy/copy_layout_stride.cpp:
```diff
@@ -113,115 +115,41 @@ void BM_MDSpan_Copy_2D_stride(benchmark::State& state, MDSpan, LayoutMapping map
-    std::array<size_t, 1>{100}
-  )
-);
-
```
Why did you remove these?

dalg24 commented on compilation_tests/ctest_constructor_sfinae.cpp:
```diff
@@ -54,92 +54,92 @@ namespace stdex = std::experimental;
 MDSPAN_STATIC_TEST(
     std::is_constructible<
-        stdex::extents<2, stdex::dynamic_extent>,
+        stdex::extents<size_t,2, stdex::dynamic_extent>,
```
don't we use clang-format?

crtrott replied:
not yet :-(

dalg24 commented on compilation_tests/ctest_trivially_copyable.cpp:
```diff
@@ -52,17 +52,10 @@ namespace stdex = std::experimental;
-);
-
-MDSPAN_STATIC_TEST(
-  !std::is_base_of<stdex::detail::__partially_static_sizes<1, 2, 3>, stdex::extents<1, 2, 3>>::value
```
Did you mean to remove?

crtrott replied:
probably shouldn't. :-)

crtrott replied:
I put it back in

dalg24 commented on tests/test_mdspan_conversion.cpp:
```diff
@@ -51,9 +51,9 @@ namespace stdex = std::experimental;
 TEST(TestMdspanConversionConst, test_mdspan_conversion_const) {
   std::array<double, 6> a{};
-  stdex::mdspan<double, stdex::extents<2, 3>> s(a.data());
+  stdex::mdspan<double, stdex::extents<uint32_t, 2, 3>> s(a.data());
```
Any particular reason here and below to chose something else than `size_t`?

crtrott replied:
To test that it actually works if it is not the default value.

dalg24 reviewed (commented):
Do you plan on adding any new test?

crtrott committed:
Fix warnings

crtrott committed:
Fix more warnings

crtrott committed:
Fix many warnings

crtrott committed:
Merge branch 'stable' into P2553-implementation

crtrott committed:
Update version

crtrott committed:
Add some configuration notes

mhoemmen commented on include/experimental/__p0009_bits/extents.hpp:
```diff
@@ -97,18 +103,20 @@ struct __extents_tag { };
+  using size_type = ThisSizeType;
 
-  using __storage_t = detail::__partially_static_sizes_tagged<detail::__extents_tag, Extents...>;
+// internal typedefs which for technical reasons are public
```
Good comment (helpful for readers)

mhoemmen commented on include/experimental/__p0009_bits/extents.hpp:
```diff
@@ -183,22 +191,24 @@ class extents
+  static constexpr rank_type rank() noexcept { return sizeof...(Extents); }
   MDSPAN_INLINE_FUNCTION
-  static constexpr size_t rank_dynamic() noexcept { return _MDSPAN_FOLD_PLUS_RIGHT((int(Extents == dynamic_extent)), /* + ... + */ 0); }
+  static constexpr rank_type rank_dynamic() noexcept { return _MDSPAN_FOLD_PLUS_RIGHT((int(Extents == dynamic_extent)), /* + ... + */ 0); }
```
Should we consider casting to `rank_type` inside, instead of to `int`, so as to avoid a potential conversion-to-unsigned warning?
```suggestion
  static constexpr rank_type rank_dynamic() noexcept { return _MDSPAN_FOLD_PLUS_RIGHT((rank_type(Extents == dynamic_extent)), /* + ... + */ 0); }
```

crtrott replied:
yeah one problem is I haven't systematically introduced rank_type yet

crtrott replied:
done

mhoemmen commented on include/experimental/__p0009_bits/extents.hpp:
```diff
@@ -183,22 +191,24 @@ class extents
   // Constructors, Destructors, and Assignment
 
   MDSPAN_INLINE_FUNCTION_DEFAULTED constexpr extents() noexcept = default;
+#if !(MDSPAN_HAS_CXX20) && false
```
This means `#if false`; should we consider just deleting this code?

crtrott replied:
jup

crtrott replied:
I remove the if for now, and then we remove the constructors later in a harmoinzation PR.

mhoemmen commented on include/experimental/__p0009_bits/layout_left.hpp:
```diff
@@ -68,20 +71,43 @@ struct layout_left {
       __rank_count<Extents::rank()-1,Extents::rank()>, const I& i) const {
       return i;
     }
+#else
```
Does this second option just exist for benchmarking?

crtrott replied:
yeah I should remove it

mhoemmen commented on include/experimental/__p0009_bits/layout_left.hpp:
```diff
@@ -68,20 +71,43 @@ struct layout_left {
+#else
+    // This option will cast __extents::extent() to the current index type.
+    // That means if you index into the mapping with just int everything gets
+    // converted to int. That means the user would need to choose index types
```
```suggestion
    // converted to int. That means users would need to choose index types
```

mhoemmen commented on include/experimental/__p0009_bits/layout_left.hpp:
```diff
@@ -68,20 +71,43 @@ struct layout_left {
+    // This option will cast __extents::extent() to the current index type.
+    // That means if you index into the mapping with just int everything gets
+    // converted to int. That means the user would need to choose index types
+    // which he knows won't overflow i.e. the index type needs to be able
```
```suggestion
    // which they know won't overflow; i.e., the index type needs to be able
```

mhoemmen commented on include/experimental/__p0009_bits/layout_left.hpp:
```diff
@@ -200,8 +225,8 @@ struct layout_left {
-    constexpr size_type __get_stride(std::experimental::extents<E...>,integer_sequence<size_t, Idx...>) const {
+    template<size_t N, class SizeType, size_t ... E, size_t ... Idx>
+    constexpr size_type __get_stride(std::experimental::extents<SizeType, E...>,integer_sequence<size_t, Idx...>) const {
       return _MDSPAN_FOLD_TIMES_RIGHT((Idx<N? __extents.template __extent<Idx>():1),1);
```
Is the type of `__extent<Idx>()` always `size_t`?

crtrott replied:
no returns `size_type`, but that should be ok or? I mean `__get_stride` also returns `size_type`?

mhoemmen commented on include/experimental/__p0009_bits/layout_right.hpp:
```diff
@@ -68,24 +71,49 @@ struct layout_right {
+#else
+    // This option will cast __extents::extent() to the current index type.
+    // That means if you index into the mapping with just int everything gets
+    // converted to int. That means the user would need to choose index types
```
```suggestion
    // converted to int. That means users would need to choose index types
```

crtrott replied:
removed this all, since it was discussion of alternative options, the first ne is what we must DO

mhoemmen commented on include/experimental/__p0009_bits/layout_right.hpp:
```diff
@@ -68,24 +71,49 @@ struct layout_right {
+    // This option will cast __extents::extent() to the current index type.
+    // That means if you index into the mapping with just int everything gets
+    // converted to int. That means the user would need to choose index types
+    // which he knows won't overflow i.e. the index type needs to be able
```
```suggestion
    // which they know won't overflow; i.e., the index type needs to be able
```

mhoemmen commented on include/experimental/__p0009_bits/maybe_static_value.hpp:
```diff
@@ -61,18 +61,18 @@ namespace experimental {
-template <class _T, _T __v,
-          _T __is_dynamic_sentinal = dynamic_extent,
+template <class _dynamic_t, class static_t, _static_t __v,
+          _static_t __is_dynamic_sentinal = dynamic_extent,
```
Just a spelling error:
```suggestion
          _static_t __is_dynamic_sentinel = dynamic_extent,
```

crtrott replied:
I will search for all of those later.

mhoemmen commented on include/experimental/__p0009_bits/maybe_static_value.hpp:
```diff
@@ -102,20 +102,20 @@ struct __maybe_static_value {
 // dynamic case
-template <class _T, _T __is_dynamic_sentinal, size_t __array_entry_index>
-struct __maybe_static_value<_T, __is_dynamic_sentinal, __is_dynamic_sentinal,
+template <class _dynamic_t, class _static_t, _static_t __is_dynamic_sentinal, size_t __array_entry_index>
```
(spelling error)
```suggestion
template <class _dynamic_t, class _static_t, _static_t __is_dynamic_sentinel, size_t __array_entry_index>
```

mhoemmen commented on include/experimental/__p0009_bits/maybe_static_value.hpp:
```diff
@@ -102,20 +102,20 @@ struct __maybe_static_value {
-template <class _T, _T __is_dynamic_sentinal, size_t __array_entry_index>
-struct __maybe_static_value<_T, __is_dynamic_sentinal, __is_dynamic_sentinal,
+template <class _dynamic_t, class _static_t, _static_t __is_dynamic_sentinal, size_t __array_entry_index>
+struct __maybe_static_value<_dynamic_t, _static_t, __is_dynamic_sentinal, __is_dynamic_sentinal,
```
```suggestion
struct __maybe_static_value<_dynamic_t, _static_t, __is_dynamic_sentinel, __is_dynamic_sentinel,
```

mhoemmen commented on include/experimental/__p0009_bits/maybe_static_value.hpp:
```diff
@@ -102,20 +102,20 @@ struct __maybe_static_value {
 #endif
 {
-  static constexpr _T __static_value = __is_dynamic_sentinal;
+  static constexpr _static_t __static_value = __is_dynamic_sentinal;
```
```suggestion
  static constexpr _static_t __static_value = __is_dynamic_sentinel;
```

mhoemmen commented on include/experimental/__p0009_bits/mdspan.hpp:
```diff
@@ -335,20 +336,20 @@ MDSPAN_TEMPLATE_REQUIRES(
-template <class ElementType, size_t... ExtentsPack>
-mdspan(ElementType*, const extents<ExtentsPack...>&)
-  -> mdspan<ElementType, ::std::experimental::extents<ExtentsPack...>>;
+template <class ElementType, class SizeType, size_t... ExtentsPack>
```
This one isn't in the proposal; do we need to add it?

crtrott replied:
no we need to remove it, since it should work on its own (i.e. all template arguments are deducable?)

crtrott replied:
but will do that as part of harmonization PR>

mhoemmen commented on include/experimental/__p0009_bits/standard_layout_static_array.hpp:
```diff
@@ -88,29 +88,29 @@ template <size_t _I, class _T> using __repeated_with_idxs = _T;
  * @tparam __sentinal
  */
-template <class _Tag, class _T, class _ValsSeq, _T __sentinal = dynamic_extent,
+template <class _Tag, class _T, class _static_t, class _ValsSeq, _static_t __sentinal = static_cast<_static_t>(dynamic_extent),
```
```suggestion
template <class _Tag, class _T, class _static_t, class _ValsSeq, _static_t __sentinel = static_cast<_static_t>(dynamic_extent),
```

mhoemmen replied:
Generally, "sentinal" -> sentinel

mhoemmen commented on include/experimental/__p0009_bits/standard_layout_static_array.hpp:
```diff
@@ -141,8 +141,8 @@ struct __standard_layout_psa<
   // it's definition.
```
```suggestion
  // its definition.
```

mhoemmen replied:
also, `__size_dynamic` is mispelled ("`__size_synamic`") in the line above.

mhoemmen reviewed (dismissed):
I did my best to review, but the innards really just need tests -- it's hard to tell with my eyeballs and brain what looks amiss.

crtrott committed:
Address review comments
